c/src/sbp.c

```diff
@@ -840,12 +651,130 @@ s8 sbp_payload_send(sbp_state_t *s, u16 msg_type, u16 sender_id, u8 len, u8 *pay
   return SBP_OK;
 }
 
-s8 sbp_message_send(sbp_state_t *s, u16 msg_type, u16 sender_id, const sbp_msg_t *msg, sbp_write_fn_t write) {
+/******************************************************************************
+ *
+ * V4 API
+ *
+ *****************************************************************************/
+
+s8 sbp_callback_register(sbp_state_t *s, sbp_msg_type_t msg_type,
+                                  sbp_decoded_callback_t cb, void *context,
+                                  sbp_msg_callbacks_node_t *node) {
+  sbp_callback_t callback;
+  callback.decoded = cb;
+  return sbp_register_callback_generic(s, msg_type, callback,
+                                       SBP_DECODED_CALLBACK, context, node);
+}
+
+s8 sbp_all_message_callback_register(sbp_state_t *s,
+                                 sbp_decoded_callback_t cb, void *context,
+                                 sbp_msg_callbacks_node_t *node) {
+  sbp_callback_t callback;
+  callback.decoded = cb;
+  return sbp_register_callback_generic(s, SBP_MSG_ALL, callback,
+                                       SBP_DECODED_CALLBACK, context, node);
+}
+
+s8 sbp_message_send(sbp_state_t *s, sbp_msg_type_t msg_type, u16 sender_id, const sbp_msg_t *msg, sbp_write_fn_t write) {
   uint8_t payload[SBP_MAX_PAYLOAD_LEN];
   uint8_t payload_len;
   s8 ret = sbp_message_encode(payload, sizeof(payload), &payload_len, msg_type, msg);
   if (ret != SBP_OK) { return ret; }
-  return sbp_payload_send(s, msg_type, sender_id, payload_len, payload, write);
+  return send_payload(s, msg_type, sender_id, payload_len, payload, write);
+}
+
+s8 sbp_process_message(sbp_state_t *s, u16 sender_id, sbp_msg_type_t msg_type,
```

Comment:
As far as I could tell, we don't have any tests for this, should we?

Reply:
Also, slightly confused, maybe I'm completely missing something but how do you process a raw frame with the new v4 API? As far as I can tell this is what `process_frame` does but nothing in the new API calls it?

Reply:
You're right, there aren't any tests for this specifically yet.

In the v4 API the idea is to never interact with encoded frames at all. `process_frame` is what's called from `sbp_process` (ie, reading data from the input stream) and also from the legacy API `sbp_frame_process`. The unfortunate duplication of the loop for processing callbacks is because `process_frame` needs to accept an encoded frame which the v4 API will only accept a decoded message as its input. They kind of work opposite from each other as far as encoding/decoding is concerned, but they process callbacks in the same way.

Reply:
Got it - thanks! I was missing that both users for the legacy and v4 APIs were able to use `sbp_process` but that makes sense.